src/pkg/state/state.go

```diff
@@ -0,0 +1,279 @@
+	"github.com/zarf-dev/zarf/src/pkg/message"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+type PkgStatus struct {
+	Name   string `json:"name"`
+	Status string `json:"status"`
+}
+
+type BundleState struct {
+	Name        string      `json:"name"`
+	PkgStatuses []PkgStatus `json:"packages"`
+	Status      string      `json:"status"`
+}
+
+type Client struct {
+	client kubernetes.Interface
+}
+
+const (
+	Success     = "success"
+	Failed      = "failed"
+	Deploying   = "deploying"
+	NotDeployed = "not_deployed" // package is in the bundle but not deployed
+	Removing    = "removing"
+	Removed     = "removed"
+	stateNs     = "uds"
+)
+
+// NewClient creates a new state client
+func NewClient(client kubernetes.Interface) (*Client, error) {
+	stateClient := &Client{
+		client: client,
+	}
+	return stateClient, nil
+}
+
+// InitBundleState initializes the bundle state in the K8s cluster if it doesn't exist.
+// This can safely be called multiple times
+func (c *Client) InitBundleState(b types.UDSBundle) error {
+	err := c.ensureNamespace()
+	if err != nil {
+		return err
+	}
+	_, err = c.getOrCreateBundleState(b)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// ensureNamespace creates the uds namespace if it doesn't exist
+func (c *Client) ensureNamespace() error {
+	_, err := c.client.CoreV1().Namespaces().Get(context.TODO(), stateNs, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			ns := &corev1.Namespace{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: stateNs,
+				},
+			}
+			_, err = c.client.CoreV1().Namespaces().Create(context.TODO(), ns, metav1.CreateOptions{})
+			if err != nil {
+				return fmt.Errorf("failed to create namespace: %w", err)
+			}
+		} else {
+			return fmt.Errorf("failed to get namespace: %w", err)
+		}
+	}
+	return nil
+}
+
+// getOrCreateBundleState gets or creates the bundle state in the K8s cluster
+func (c *Client) getOrCreateBundleState(b types.UDSBundle) (*BundleState, error) {
+	var state *BundleState
+	bundleName := b.Metadata.Name
+	stateSecretName := fmt.Sprintf("uds-bundle-%s", bundleName)
+	stateSecret, err := c.client.CoreV1().Secrets(stateNs).Get(context.TODO(), stateSecretName, metav1.GetOptions{})
+	if err != nil {
+		if errors.IsNotFound(err) {
+			var pkgStatuses []PkgStatus
+			for _, pkg := range b.Packages {
+				pkgStatuses = append(pkgStatuses, PkgStatus{Name: pkg.Name, Status: NotDeployed})
+			}
+
+			// init state and secret
+			state = &BundleState{
+				Name:        bundleName,
+				PkgStatuses: pkgStatuses,
+			}
+
+			stateSecret = &corev1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: stateSecretName,
+				},
+				Data: map[string][]byte{
+					"data": {},
+				},
+			}
+			_, err = c.client.CoreV1().Secrets(stateNs).Create(context.TODO(), stateSecret, metav1.CreateOptions{})
+			if err != nil {
+				return nil, err
+			}
+		}
```

Comment:
Is it possible to get another error here that is not a "IsNotFound"?

Reply:
yes updated